fd_readwrite_lock.h is not protected by FD_HAS_ATOMIC feature

The runtime fails to build for targets without threads:

```
src/flamenco/runtime/fd_readwrite_lock.h:49:31: error: call to undeclared function 'FD_ATOMIC_CAS'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
   49 |     if( FD_UNLIKELY( l < 0 || FD_ATOMIC_CAS( &lock->lock, l, l+1 ) != l ) ) {
      |                               ^
src/flamenco/runtime/fd_readwrite_lock.h:66:22: error: call to undeclared function 'FD_ATOMIC_CAS'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
   66 |     if( FD_UNLIKELY( FD_ATOMIC_CAS( &lock->lock, l, l-1 ) != l ) ) {
      |                      ^
src/flamenco/runtime/fd_readwrite_lock.h:78:32: error: call to undeclared function 'FD_ATOMIC_CAS'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
   78 |     if( FD_UNLIKELY( l != 0 || FD_ATOMIC_CAS( &lock->lock, l, -1 ) != l ) ) {
      |                                ^
src/flamenco/runtime/fd_readwrite_lock.h:99:22: error: call to undeclared function 'FD_ATOMIC_CAS'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
   99 |     if( FD_UNLIKELY( FD_ATOMIC_CAS( &lock->lock, l, 0 ) != l ) ) {
      |                      ^
4 errors generated.
```

--- a/src/app/fdctl/Local.mk
+++ b/src/app/fdctl/Local.mk
@@ -1,4 +1,5 @@
 ifdef FD_HAS_HOSTED
+ifdef FD_HAS_THREADS
 ifdef FD_HAS_ALLOCA
 ifdef FD_HAS_DOUBLE
 ifdef FD_HAS_INT128
@@ -151,6 +152,7 @@ $(OBJDIR)/bin/solana: solana/target/$(RUST_PROFILE)/solana
 
 solana: $(OBJDIR)/bin/solana $(OBJDIR)/bin/solana
 
+endif
 endif
 endif
 endif

--- a/src/disco/bank/fd_rwlock.h
+++ b/src/disco/bank/fd_rwlock.h
@@ -18,6 +18,7 @@ typedef struct fd_rwlock fd_rwlock_t;
 
 static inline void
 fd_rwlock_write( fd_rwlock_t * lock ) {
+# if FD_HAS_THREADS
   for(;;) {
     ushort value = lock->value;
     if( FD_LIKELY( !value ) ) {
@@ -25,6 +26,9 @@ fd_rwlock_write( fd_rwlock_t * lock ) {
     }
     FD_SPIN_PAUSE();
   }
+# else
+  lock->value = 0xFFFF;
+# endif
   FD_COMPILER_MFENCE();
 }
 
@@ -36,6 +40,7 @@ fd_rwlock_unwrite( fd_rwlock_t * lock ) {
 
 static inline void
 fd_rwlock_read( fd_rwlock_t * lock ) {
+# if FD_HAS_THREADS
   for(;;) {
     ushort value = lock->value;
     if( FD_UNLIKELY( value!=0xFFFE ) ) {
@@ -45,13 +50,20 @@ fd_rwlock_read( fd_rwlock_t * lock ) {
     }
     FD_SPIN_PAUSE();
   }
+# else
+  lock->value++;
+# endif
   FD_COMPILER_MFENCE();
 }
 
 static inline void
 fd_rwlock_unread( fd_rwlock_t * lock ) {
   FD_COMPILER_MFENCE();
+# if FD_HAS_THREADS
   FD_ATOMIC_FETCH_AND_SUB( &lock->value, 1 );
+# else
+  lock->value--;
+# endif
 }
 
 #endif /* HEADER_fd_src_disco_bank_rwlock_h */

--- a/src/disco/topo/Local.mk
+++ b/src/disco/topo/Local.mk
@@ -1,6 +1,8 @@
 ifdef FD_HAS_HOSTED
+ifdef FD_HAS_THREADS
 ifdef FD_HAS_LINUX
 $(call add-hdrs,fd_topo.h fd_pod_format.h)
 $(call add-objs,fd_topo fd_topob fd_topo_run,fd_disco)
+endif
 endif
 endif

--- a/src/disco/tvu/fd_pending_slots.c
+++ b/src/disco/tvu/fd_pending_slots.c
@@ -25,7 +25,7 @@ fd_pending_slots_new( void * mem, ulong lo_wmark ) {
 
   laddr += sizeof( fd_pending_slots_t );
   pending_slots->pending = (void *)laddr;
-  
+
   laddr += sizeof(long) * FD_PENDING_MAX;
 
   FD_TEST( laddr == (ulong)mem + footprint );
@@ -78,10 +78,14 @@ fd_pending_slots_delete( void * pending_slots ) {
 
 static void
 fd_pending_slots_lock( fd_pending_slots_t * pending_slots ) {
+# if FD_HAS_THREADS
   for(;;) {
     if( FD_LIKELY( !FD_ATOMIC_CAS( &pending_slots->lock, 0UL, 1UL) ) ) break;
     FD_SPIN_PAUSE();
   }
+# else
+  pending_slots->lock = 1UL;
+# endif
   FD_COMPILER_MFENCE();
 }
 
@@ -134,14 +138,14 @@ fd_pending_slots_add( fd_pending_slots_t * pending_slots,
                       ulong slot,
                       long when ) {
   fd_pending_slots_lock( pending_slots );
-  
+
   long * pending = pending_slots->pending;
   if( pending_slots->start == pending_slots->end ) {
     /* Queue is empty */
     pending_slots->start = slot;
     pending_slots->end = slot+1U;
     pending[slot & FD_PENDING_MASK] = when;
-    
+
   } else if ( slot < pending_slots->start ) {
     /* Grow down */
     if( (long)(pending_slots->end - slot) > (long)FD_PENDING_MAX )

--- a/src/disco/tvu/fd_replay.c
+++ b/src/disco/tvu/fd_replay.c
@@ -102,10 +102,14 @@ fd_replay_delete( void * replay ) {
 
 static void
 fd_replay_pending_lock( fd_replay_t * replay ) {
+# if FD_HAS_THREADS
   for( ;; ) {
     if( FD_LIKELY( !FD_ATOMIC_CAS( &replay->pending_lock, 0UL, 1UL ) ) ) break;
     FD_SPIN_PAUSE();
   }
+# else
+  replay->pending_lock = 1;
+# endif
   FD_COMPILER_MFENCE();
 }
 

--- a/src/flamenco/gossip/fd_gossip.c
+++ b/src/flamenco/gossip/fd_gossip.c
@@ -401,10 +401,14 @@ fd_gossip_delete ( void * shmap ) {
 
 static void
 fd_gossip_lock( fd_gossip_t * gossip ) {
+# if FD_HAS_THREADS
   for(;;) {
     if( FD_LIKELY( !FD_ATOMIC_CAS( &gossip->lock, 0UL, 1UL) ) ) break;
     FD_SPIN_PAUSE();
   }
+# else
+  gossip->lock = 1;
+# endif
   FD_COMPILER_MFENCE();
 }
 
@@ -511,7 +515,7 @@ fd_gossip_update_tvu_addr( fd_gossip_t * glob, const fd_gossip_peer_addr_t * tvu
 }
 
 int
-fd_gossip_update_tpu_addr( fd_gossip_t * glob, 
+fd_gossip_update_tpu_addr( fd_gossip_t * glob,
                            fd_gossip_peer_addr_t const * tpu,
                            fd_gossip_peer_addr_t const * tpu_fwd ) {
   char tmp[100];

--- a/src/flamenco/repair/fd_repair.c
+++ b/src/flamenco/repair/fd_repair.c
@@ -294,10 +294,14 @@ fd_repair_delete ( void * shmap ) {
 
 static void
 fd_repair_lock( fd_repair_t * repair ) {
+# if FD_HAS_THREADS
   for(;;) {
     if( FD_LIKELY( !FD_ATOMIC_CAS( &repair->lock, 0UL, 1UL) ) ) break;
     FD_SPIN_PAUSE();
   }
+# else
+  repair->lock = 1;
+# endif
   FD_COMPILER_MFENCE();
 }
 
@@ -695,7 +699,7 @@ fd_actives_shuffle( fd_repair_t * repair ) {
       leftovers = fd_scratch_alloc(
         alignof( fd_active_elem_t * ),
         sizeof( fd_active_elem_t * ) * fd_active_table_key_cnt( repair->actives ) );
-      
+
       for( fd_active_table_iter_t iter = fd_active_table_iter_init( repair->actives );
          !fd_active_table_iter_done( repair->actives, iter );
          iter = fd_active_table_iter_next( repair->actives, iter ) ) {
@@ -726,7 +730,7 @@ fd_actives_shuffle( fd_repair_t * repair ) {
 
     long  latencies[ FD_REPAIR_STICKY_MAX ];
     ulong latencies_cnt = 0UL;
-    
+
     long first_quartile_latency = LONG_MAX;
 
     /* fetch all latencies */

--- a/src/flamenco/runtime/fd_bank_hash_cmp.c
+++ b/src/flamenco/runtime/fd_bank_hash_cmp.c
@@ -85,10 +85,14 @@ fd_bank_hash_cmp_delete( void * bank_hash_cmp ) {
 void
 fd_bank_hash_cmp_lock( fd_bank_hash_cmp_t * bank_hash_cmp ) {
   volatile int * lock = &bank_hash_cmp->lock;
+# if FD_HAS_THREADS
   for( ;; ) {
     if( FD_LIKELY( !FD_ATOMIC_CAS( lock, 0UL, 1UL ) ) ) break;
     FD_SPIN_PAUSE();
   }
+# else
+  *lock = 1;
+# endif
   FD_COMPILER_MFENCE();
 }
 

--- a/src/flamenco/runtime/fd_readwrite_lock.h
+++ b/src/flamenco/runtime/fd_readwrite_lock.h
@@ -44,6 +44,7 @@ fd_readwrite_delete( void * shreadwrite ) {
 
 static inline void
 fd_readwrite_start_read( fd_readwrite_lock_t * lock ) {
+# if FD_HAS_THREADS
   for(;;) {
     register int l = lock->lock;
     if( FD_UNLIKELY( l < 0 || FD_ATOMIC_CAS( &lock->lock, l, l+1 ) != l ) ) {
@@ -53,10 +54,14 @@ fd_readwrite_start_read( fd_readwrite_lock_t * lock ) {
     break;
   }
   FD_COMPILER_MFENCE();
+# else
+  lock->lock++;
+# endif
 }
 
 static inline void
 fd_readwrite_end_read( fd_readwrite_lock_t * lock ) {
+# if FD_HAS_THREADS
   FD_COMPILER_MFENCE();
   for(;;) {
     register int l = lock->lock;
@@ -69,10 +74,14 @@ fd_readwrite_end_read( fd_readwrite_lock_t * lock ) {
     }
     break;
   }
+# else
+  lock->lock--;
+# endif
 }
 
 static inline void
 fd_readwrite_start_write( fd_readwrite_lock_t * lock ) {
+# if FD_HAS_THREADS
   for(;;) {
     register int l = lock->lock;
     if( FD_UNLIKELY( l != 0 || FD_ATOMIC_CAS( &lock->lock, l, -1 ) != l ) ) {
@@ -81,6 +90,9 @@ fd_readwrite_start_write( fd_readwrite_lock_t * lock ) {
     }
     break;
   }
+# else
+  lock->lock = -1;
+# endif
   FD_COMPILER_MFENCE();
   lock->seqnum++;
   FD_COMPILER_MFENCE();
@@ -91,6 +103,7 @@ fd_readwrite_end_write( fd_readwrite_lock_t * lock ) {
   FD_COMPILER_MFENCE();
   lock->seqnum++;
   FD_COMPILER_MFENCE();
+# if FD_HAS_THREADS
   for(;;) {
     register int l = lock->lock;
     if( FD_UNLIKELY( l >= 0 ) ) {
@@ -102,6 +115,9 @@ fd_readwrite_end_write( fd_readwrite_lock_t * lock ) {
     }
     break;
   }
+# else
+  lock->lock = 0;
+# endif
 }
 
 /* The pattern for concurrent reads is:

--- a/src/flamenco/snapshot/Local.mk
+++ b/src/flamenco/snapshot/Local.mk
@@ -7,7 +7,9 @@ $(call add-hdrs,fd_snapshot_http.h)
 $(call add-objs,fd_snapshot_http,fd_flamenco)
 $(call make-unit-test,test_snapshot_http,test_snapshot_http,fd_flamenco fd_disco fd_funk fd_ballet fd_util)
 $(call run-unit-test,test_snapshot_http)
+ifdef FD_HAS_THREADS
 $(call make-fuzz-test,fuzz_snapshot_http,fuzz_snapshot_http,fd_flamenco fd_disco fd_funk fd_ballet fd_util)
+endif
 endif
 
 $(call add-hdrs,fd_snapshot_istream.h)

